### PR TITLE
big revamp

### PR DESCRIPTION
**Pull request by trishorts**

### CODE REVIEW

File: EngineLayer/CompactPeptideWithModifiedMass.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Proteomics.Fragmentation;
+using Proteomics.ProteolyticDigestion;
+
+namespace EngineLayer
+{
+    //[Serializable]
```
Comment (trishorts) [inline EngineLayer/CompactPeptideWithModifiedMass.cs]: Zach, do we need this block of code for anything?

File: EngineLayer/CompactPeptideWithModifiedMass.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    [Serializable]
+    public class PeptideWithSetModifications_ModifiedMass
+    {
+        public double[] CterminalMasses { get; private set; }
+        public double[] NterminalMasses { get; private set; }
+        public double PeptideWithSetModifications_MonoisotopicMass { get; private set; }
+        public double ModifiedMass { get; private set; }
+
+
+        public PeptideWithSetModifications_ModifiedMass(PeptideWithSetModifications p, double newMonoMass, MassSpectrometry.DissociationType d)
```
Comment (trishorts) [inline EngineLayer/CompactPeptideWithModifiedMass.cs]: This is unreferenced. Do we need it?

File: EngineLayer/CrosslinkSearch/TwoPassCrosslinkSearchEngine.cs [hunk 3]
```diff
@@ -9,53 +12,53 @@
...
+        private readonly Crosslinker Crosslinker;
+
+        private readonly bool CrosslinkSearchTopN;
+        private readonly int TopN;
         private readonly Tolerance XLPrecusorMsTl;
         private readonly bool QuenchH2O;
         private readonly bool QuenchNH2;
         private readonly bool QuenchTris;
         private readonly bool Charge_2_3;
+        private readonly bool Charge_2_3_PrimeFragment;
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/TwoPassCrosslinkSearchEngine.cs]: some comments explaining these bools would be helpful

File: EngineLayer/CompactPeptideWithModifiedMass.cs (hunk 1 above)
Comment (zrolfs) [reply EngineLayer/CompactPeptideWithModifiedMass.cs]: No, delete it.

File: EngineLayer/CompactPeptideWithModifiedMass.cs (hunk 2 above)
Comment (zrolfs) [reply EngineLayer/CompactPeptideWithModifiedMass.cs]: I don't think this class is needed anymore. It previously existed for converting compact peptides into pwsms.

File: EngineLayer/CrosslinkSearch/TwoPassCrosslinkSearchEngine.cs [hunk 4]
```diff
@@ -147,285 +167,328 @@ protected override MetaMorpheusEngineResults RunSpecific()
...
-                            PsmCross psmCross = null;
-                            if (Crosslinker.CrosslinkerModSites == Crosslinker.CrosslinkerModSites2)
-                            {
-                                psmCross = XlSitesSame(theScan, psmCrossAlpha, psmCrossBeta, Crosslinker, ind, inx, x, y);
-                            }
-                            else
+                            if (!possibleCrosslinkLocationsBeta.Any())
                             {
-                                psmCross = XlSitesTwoDiff(theScan, psmCrossAlpha, psmCrossBeta, Crosslinker, ind, inx, x, y);
+                                continue;
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/TwoPassCrosslinkSearchEngine.cs]: why are we doing this?

File: EngineLayer/ModernSearch/ModernSearchEngine.cs [hunk 5]
```diff
@@ -191,38 +207,37 @@ protected List<int> GetBinsToSearch(Ms2ScanWithSpecificMass scan)
...
-                        {
-                            protonMassShift = ClassExtensions.ToMass(protonMassShift, 1);
-                            int compFragmentFloorMass = (int)Math.Round(((scan.PrecursorMass + protonMassShift) * FragmentBinsPerDalton)) - obsFragmentCeilingMass;
-                            int compFragmentCeilingMass = (int)Math.Round(((scan.PrecursorMass + protonMassShift) * FragmentBinsPerDalton)) - obsFragmentFloorMass;
 
-                            // prevent index out of bounds errors
-                            if (compFragmentCeilingMass >= FragmentIndex.Length)
-                            {
-                                compFragmentCeilingMass = FragmentIndex.Length - 1;
+                    if (complementaryIonConversionDictionary.TryGetValue(commonParameters.DissociationType, out double protonMassShift)) //TODO: this is broken for EThcD because that method needs two conversions
```
Comment (trishorts) [inline EngineLayer/ModernSearch/ModernSearchEngine.cs]: do we have unit tests that cover this code block?

Comment (zrolfs) [reply EngineLayer/ModernSearch/ModernSearchEngine.cs]: The non-specific tests should be covering this.

File: EngineLayer/Mods/fallOff.txt [hunk 6]
```diff
@@ -1,253 +1,260 @@
...
-//
-ID   fallOffN L
-PP   Peptide N-terminal.
-TG   L
-MT   fallOffN
-CF   C-6H-11N-1O-1
-//
-ID   fallOffN C
-PP   Peptide N-terminal.
+Glycosylations
```
Comment (trishorts) [inline EngineLayer/Mods/fallOff.txt]: I'd like for us to discuss what we do with glyco mods and how we deal with their neutral losses

File: EngineLayer/NonSpecificEnzymeSearch/NonSpecificEnzymeSearchEngine.cs [hunk 7]
```diff
@@ -145,63 +229,203 @@ protected override MetaMorpheusEngineResults RunSpecific()
             return new MetaMorpheusEngineResults(this);
         }
 
-        private Tuple<int, double> Accepts(double scanPrecursorMass, CompactPeptide peptide, TerminusType terminusType, MassDiffAcceptor searchMode)
+        //public static IEnumerable<Product> Fragment( DissociationType dissociationType, FragmentationTerminus fragmentationTerminus, PeptideWithSetModifications pwsm)
```
Comment (trishorts) [inline EngineLayer/NonSpecificEnzymeSearch/NonSpecificEnzymeSearchEngine.cs]: I think this was here for debugging purposes and should be deleted. Zach?

Comment (zrolfs) [reply EngineLayer/NonSpecificEnzymeSearch/NonSpecificEnzymeSearchEngine.cs]: Yes, delete

File: GUI/CalibrateTaskWindow.xaml [hunk 8]
```diff
@@ -5,8 +5,20 @@
         xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
         xmlns:local="clr-namespace:MetaMorpheusGUI"
         mc:Ignorable="d"
+        xmlns:sys="clr-namespace:System;assembly=mscorlib"
+        xmlns:mzl="clr-namespace:MassSpectrometry;assembly=MassSpectrometry"
         KeyDown="KeyPressed"
         Title="Calibrate Task" Width="750" Height="750"  WindowStartupLocation="CenterScreen">
+
```
Comment (trishorts) [inline GUI/CalibrateTaskWindow.xaml]: I think the positioning of the pull-down menu needs to be adjusted. Also make sure we don't offer custom dissociation yet.

File: GUI/MetaDraw.xaml.cs [hunk 9]
```diff
@@ -250,31 +251,59 @@ private void DrawAnnotatedBaseSequence(MetaDrawPsm psm)
             }
 
             // draw b ions
-            foreach (var bIon in psm.FragmentIons.Where(p => p.ProductType == ProductType.B))
+            foreach (var bIon in psm.FragmentIons.Where(p => p.NeutralTheoreticalProduct.ProductType == ProductType.b))
```
Comment (trishorts) [inline GUI/MetaDraw.xaml.cs]: this old-style looping through the 4 accepted ion types seems out of date in the new era

File: GUI/MetaDraw/PsmAnnotationViewModel.cs [hunk 10]
```diff
@@ -20,11 +22,10 @@ public class PsmAnnotationViewModel : INotifyPropertyChanged
         private PlotModel privateModel;
 
         private static Dictionary<ProductType, OxyColor> productTypeDrawColors = new Dictionary<ProductType, OxyColor>
-        { { ProductType.B, OxyColors.Blue },
-          { ProductType.BnoB1ions, OxyColors.Blue },
-          { ProductType.Y, OxyColors.Purple },
-          { ProductType.C, OxyColors.Gold },
-          { ProductType.Zdot, OxyColors.Orange } };
+        { { ProductType.b, OxyColors.Blue },
```
Comment (trishorts) [inline GUI/MetaDraw/PsmAnnotationViewModel.cs]: let's use the complete sets of product types from mzlib

File: TaskLayer/GPTMDTask/GPTMDParameters.cs [hunk 11]
```diff
@@ -8,14 +8,14 @@ public class GptmdParameters
...
             ListOfModsGptmd = GlobalVariables.AllModsKnown.Where(b =>
-                b.modificationType.Equals("N-linked glycosylation") ||
-                b.modificationType.Equals("Other glycosylation") ||
-                b.modificationType.Equals("Mod") ||
-                //b.modificationType.Equals("Artifact") ||
-                //b.modificationType.Equals("Biological") ||
-                b.modificationType.Equals("PeptideTermMod") ||
-                b.modificationType.Equals("Metal") ||
-                b.modificationType.Equals("ProteinTermMod")).Select(b => (b.modificationType, b.id)).ToList();
+                b.ModificationType.Equals("N-linked glycosylation") ||
```
Comment (trishorts) [inline TaskLayer/GPTMDTask/GPTMDParameters.cs]: guess we need to decide what we're going to offer and what is available in the lists

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 12]
```diff
@@ -84,169 +68,185 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
-                    }
+                    WriteAllToTsv(allPsms, OutputFolder, "allPsms", new List<string> { taskId });
                 }
-
+                catch (Exception)
                 {
-                    var interPsmsXL = allPsmsXL.Where(p => !p.CompactPeptides.First().Value.Item2.Select(b => b.Protein.Accession).First().Contains(p.BetaPsmCross.CompactPeptides.First().Value.Item2.Select(b => b.Protein.Accession).First()) &&
-                        !p.BetaPsmCross.CompactPeptides.First().Value.Item2.Select(b => b.Protein.Accession).First().Contains(p.CompactPeptides.First().Value.Item2.Select(b => b.Protein.Accession).First())).OrderByDescending(p => p.XLQvalueTotalScore).Select(p => { p.CrossType = PsmCrossType.Inter; return p; }).ToList();
-                    var interPsmsXLFDR = CrosslinkDoFalseDiscoveryRateAnalysis(interPsmsXL).ToList();
+                    throw;
```
Comment (trishorts) [inline TaskLayer/XLSearchTask/XLSearchTask.cs]: how about a message?

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 13]
```diff
@@ -84,169 +68,185 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
+            if (XlSearchParameters.XlOutPercolator)
+            {
+                try
+                {
+                    var interPsmsXLPercolator = interPsmsXL.Where(p => p.BestScore >= 2 && p.BetaPeptide.BestScore >= 2).OrderBy(p => p.ScanNumber).ToList();
+                    WriteCrosslinkToTxtForPercolator(interPsmsXLPercolator, OutputFolder, "xl_inter_perc", crosslinker, new List<string> { taskId });
+                }
+                catch (Exception)
+                {
+                    throw;
```
Comment (trishorts) [inline TaskLayer/XLSearchTask/XLSearchTask.cs]: message

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 14]
```diff
@@ -84,169 +68,185 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
+            if (XlSearchParameters.XlOutPercolator)
+            {
+                try
+                {
+                    var intraPsmsXLPercolator = intraPsmsXL.Where(p => p.BestScore >= 2 && p.BetaPeptide.BestScore >= 2).OrderBy(p => p.ScanNumber).ToList();
+                    WriteCrosslinkToTxtForPercolator(intraPsmsXLPercolator, OutputFolder, "xl_intra_perc", crosslinker, new List<string> { taskId });
+                }
+                catch (Exception)
+                {
+                    throw;
```
Comment (trishorts) [inline TaskLayer/XLSearchTask/XLSearchTask.cs]: message

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 15]
```diff
@@ -84,169 +68,185 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
+                allPsmsXLFDR.AddRange(interPsmsXLFDR.Where(p => p.IsDecoy != true && p.BetaPeptide.IsDecoy != true && p.FdrInfo.QValue <= 0.05).ToList());
+                try
+                {
+                    allPsmsXLFDR = allPsmsXLFDR.OrderByDescending(p => p.XLQvalueTotalScore).ToList();
+                    var allPsmsXLFDRGroup = FindCrosslinks(allPsmsXLFDR);
+                    WritePsmCrossToTsv(allPsmsXLFDRGroup, Path.Combine(OutputFolder, "allPsmsXLFDRGroup.tsv"), 2);
+                }
+                catch (Exception)
+                {
+                    throw;
```
Comment (trishorts) [inline TaskLayer/XLSearchTask/XLSearchTask.cs]: message

File: Test/AddCompIonsTest.cs [hunk 16]
```diff
@@ -151,10 +152,10 @@ public static void TestCompIons_MatchIonsScore()
...
-            List<ProductType> lp = new List<ProductType> { ProductType.B, ProductType.Y };
-            double scoreT = MetaMorpheusEngine.CalculatePeptideScoreOld(t.GetOneBasedScan(2), productMassTolerance, sorted_theoretical_product_masses_for_this_peptide, precursorMass, new List<DissociationType> { DissociationType.HCD }, true, 0);
-            double scoreF = MetaMorpheusEngine.CalculatePeptideScoreOld(t.GetOneBasedScan(2), productMassTolerance, sorted_theoretical_product_masses_for_this_peptide, precursorMass, new List<DissociationType> { DissociationType.HCD }, false, 0);
-            Assert.IsTrue(scoreT == scoreF * 2 && scoreT > scoreF + 1);
+            List<ProductType> lp = new List<ProductType> { ProductType.b, ProductType.y };
+            //double scoreT = MetaMorpheusEngine.CalculatePeptideScore(t.GetOneBasedScan(2), productMassTolerance, DissociationType.HCD, sorted_theoretical_product_masses_for_this_peptide, precursorMass, true, 0);
```
Comment (trishorts) [inline Test/AddCompIonsTest.cs]: should I stay or should I go?

File: Test/MultiProteaseParsimonyTest.cs [hunk 17]
```diff
@@ -666,136 +448,47 @@ public static void MultiProteaseIndistiguishableTest()
...
-                }
-                foreach (var peptide in protein.Digest(digestionParams2, new List<ModificationWithMass>(), new List<ModificationWithMass>()))
+                foreach (var peptide in protein.Digest(digestionParams, new List<Modification>(), new List<Modification>()))
                 {
                     switch (peptide.BaseSequence)
                     {
-
                         case "ABC": peptideList.Add(peptide); break;
-                        case "EFG": peptideList.Add(peptide); break;
+
```
Comment (trishorts) [inline Test/MultiProteaseParsimonyTest.cs]: switch has only one case

File: Test/TestPsm.cs [hunk 18]
```diff
@@ -28,29 +29,26 @@ public static void TestPsmHeader()
...
-
-            psm.MatchToProteinLinkedPeptides(matching);
-
             Assert.AreEqual(psm.ToString().Count(f => f == '\t'), PeptideSpectralMatch.GetTabSeparatedHeader().Count(f => f == '\t'));
 
             Tolerance fragmentTolerance = new PpmTolerance(10);
-            List<ProductType> lp = new List<ProductType> { ProductType.B };
+            List<ProductType> lp = new List<ProductType> { ProductType.b };
 
             new LocalizationEngine(new List<PeptideSpectralMatch> { psm }, lp, myMsDataFile, new CommonParameters(productMassTolerance: fragmentTolerance), new List<string>()).Run();
```
Comment (trishorts) [inline Test/TestPsm.cs]: localization engine can get product types from common params etc.

File: Test/AddCompIonsTest.cs (hunk 16 above)
Comment (zrolfs) [reply Test/AddCompIonsTest.cs]: [Test]
        public static void TestCompIons_MatchIonsScore()
        {
            TestDataFile t = new TestDataFile();
            Tolerance productMassTolerance = new AbsoluteTolerance(0.01);
            double precursorMass = 300;
            //The below theoretical does not accurately represent B-Y ions
            double[] sorted_theoretical_product_masses_for_this_peptide = new double[] { precursorMass + (2 * Constants.ProtonMass) - 275.1350, precursorMass + (2 * Constants.ProtonMass) - 258.127, precursorMass + (2 * Constants.ProtonMass) - 257.1244, 50, 60, 70, 147.0764, precursorMass + (2 * Constants.ProtonMass) - 147.0764, precursorMass + (2 * Constants.ProtonMass) - 70, precursorMass + (2 * Constants.ProtonMass) - 60, precursorMass + (2 * Constants.ProtonMass) - 50, 257.1244, 258.127, 275.1350 }; //{ 50, 60, 70, 147.0764, 257.1244, 258.127, 275.1350 }
            List<Product> productsWithLocalizedMassDiff = new List<Product>();
            foreach (double d in sorted_theoretical_product_masses_for_this_peptide)
            {
                NeutralTerminusFragment frag = new NeutralTerminusFragment(FragmentationTerminus.Both, d, 1, 1);
                productsWithLocalizedMassDiff.Add(new Product(ProductType.b, frag, 0));
            }
            CommonParameters commonParameters = new CommonParameters { ProductMassTolerance = new AbsoluteTolerance(0.01) };

            MsDataScan scan = t.GetOneBasedScan(2);
            List<MatchedFragmentIon> matchedIons = MetaMorpheusEngine.MatchFragmentIons(scan.MassSpectrum, productsWithLocalizedMassDiff, commonParameters);


            MzSpectrum complementarySpectrum = MetaMorpheusEngine.GenerateComplementarySpectrum(scan.MassSpectrum, precursorMass, commonParameters.DissociationType);
            List<MatchedFragmentIon> matchedCompIons = MetaMorpheusEngine.MatchFragmentIons(complementarySpectrum, productsWithLocalizedMassDiff, commonParameters);
            matchedCompIons.AddRange(matchedIons);

            // score when the mass-diff is on this residue
            double localizedScore = MetaMorpheusEngine.CalculatePeptideScore(scan, matchedIons, 0);
            double scoreNormal = MetaMorpheusEngine.CalculatePeptideScore(scan, matchedIons, 0);
            double scoreComp = MetaMorpheusEngine.CalculatePeptideScore(scan, matchedCompIons, 0);
            Assert.IsTrue(scoreNormal *2 == scoreComp && scoreComp > scoreNormal + 1);
        }

Comment (zrolfs) [reply Test/AddCompIonsTest.cs]: git's being weird, so "edit" that comment to copy the code

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 19]
```diff
@@ -11,358 +12,384 @@ namespace EngineLayer
...
                 }
+                else
+                {
+                    proteinToPepSeqMatch.Add(peptide.Protein, new HashSet<string> { sequence });
+                }
+            }
 
-                Protein bestProtein = possibleBestProteinList.First().Key;
-                HashSet<string> newSeqs = new HashSet<string>(algDictionary[bestProtein]);
+            // remove the peptides observed by proteins with unique peptides
```
Comment (rmmiller22) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: In Multiprotease Parsimony there is an bug being introduced here. I am working to find a solution.

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 20]
```diff
@@ -11,358 +12,384 @@ namespace EngineLayer
...
                                 {
-                                    if (parsimonyProteinWithThisNumPeptides.Key != list[i].Key
-                                    && proteinToPeptidesMatching[parsimonyProteinWithThisNumPeptides.Key].SetEquals(proteinToPeptidesMatching[list[i].Key]))
+                                    Protein otherProtein = list[i].Key;
+
+                                    foreach (var parsimonyProtein in parsimonyProteinsWithSameNumPeptides)
                                     {
-                                        indistinguishableProteins.GetOrAdd(list[i].Key, proteinToPeptidesMatching[list[i].Key]);
+                                        // if the two proteins have the same set of peptide sequences, they're indistinguishable
+                                        // TODO: consider multiprotease here?
```
Comment (rmmiller22) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: We do need to handle multiprotease conditions here. Working to figure this out.

File: EngineLayer/Mods/Mods.txt [hunk 21]
```diff
@@ -72,7 +72,7 @@ MT   Mod
 CF   O2
 //
 ID   Oxidation
-TG   M or Y or W or P or F or D or E or V or H or K or N or R or C or U or I or L or Q or S or T
+TG   Y or W or P or F or D or E or V or H or K or N or R or C or U or I or L or Q or S or T
```
Comment (acesnik) [inline EngineLayer/Mods/Mods.txt]: Wow, that's subtle.

File: TaskLayer/SearchTask/PostSearchAnalysisTask.cs [hunk 22]
```diff
@@ -74,61 +80,22 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
         }
```
Comment (zrolfs) [inline TaskLayer/SearchTask/PostSearchAnalysisTask.cs]: Should this inherit from the engine if there's no "RunSpecific"?

File: TaskLayer/SearchTask/PostSearchAnalysisTask.cs [hunk 23]
```diff
@@ -74,61 +80,22 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
 
         /// <summary>
-        /// Constructs protein groups; calculates false discovery rate (FDR) for PSMs and protein groups
+        /// Calculate estimated false-discovery rate (FDR) for peptide spectral matches (PSMs)
         /// </summary>
-        private void ProteinAndFdrAnalysis()
+        private void CalculatePsmFdr()
```
Comment (zrolfs) [inline TaskLayer/SearchTask/PostSearchAnalysisTask.cs]: This seems like an unnecessary method where instead the FDRAnalysisEngine.Run() could be used instead.

File: TaskLayer/SearchTask/PostSearchAnalysisTask.cs [hunk 24]
```diff
@@ -74,61 +80,22 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
-                    .ThenBy(b => b.PeptideMonisotopicMass.HasValue ? Math.Abs(b.ScanPrecursorMass - b.PeptideMonisotopicMass.Value) : double.MaxValue)
-                    .GroupBy(b => new Tuple<string, int, double?>(b.FullFilePath, b.ScanNumber, b.PeptideMonisotopicMass)).Select(b => b.First()).ToList();
-            }
+            // sort the list of psms by the score (for FDR analysis)
 
             // do protein parsimony and protein scoring/FDR
-            ProteinParsimonyResults proteinAnalysisResults = null;
             if (Parameters.SearchParameters.DoParsimony)
             {
                 Status("Running protein analysis...", Parameters.SearchTaskId);
```
[the remark this answers is not in the document]
Comment (rmmiller22) [reply TaskLayer/SearchTask/PostSearchAnalysisTask.cs]: In revamp the FDR filtering is performed in ProteinParsimonyEngine. It is still being done, just in a better place. There was no need for is to be in the PostSearchAnalysisTask itself.

**Pull request merged**